### PR TITLE
DROOLS-6564 DMN Java invocations returning Arrays FEEL handling

### PR DESCRIPTION
**Pull request by tarilabs**
**JIRA**: 

* https://issues.redhat.com/browse/DROOLS-6564
* https://issues.redhat.com/browse/DROOLS-6563

**referenced Pull Requests**: 

* https://github.com/kiegroup/drools/pull/3798
* https://github.com/kiegroup/droolsjbpm-integration/pull/2579

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by tarilabs** [comment]
Jenkins run cdb

**Comment by tarilabs** [comment]
```
15:00:20  [SpotBugs] Successfully parsed file /home/jenkins/workspace/KIE/main/pullrequest/drools-main.pr/bc/kiegroup_drools/kie-test-util/target/spotbugsXml.xml
15:00:20  [SpotBugs] -> found 0 issues (skipped 0 duplicates)
15:00:34  Body did not finish within grace period; terminating with extreme prejudice
[Pipeline] }
[Pipeline] // script
Error when executing always post condition:
org.jenkinsci.plugins.workflow.steps.FlowInterruptedException
	at org.jenkinsci.plugins.workflow.steps.TimeoutStepExecution.cancel(TimeoutStepExecution.java:165)
	at org.jenkinsci.plugins.workflow.steps.TimeoutStepExecution.access$100(TimeoutStepExecution.java:38)
```

🤦 

Jenkins retest this

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java [hunk 1]
```diff
@@ -362,7 +362,15 @@ private void adjustForVariableParameters(CandidateMethod cm, Class<?>[] paramete
...
         // this is to normalize types returned by external functions
-        return result != null && result instanceof Number && !(result instanceof BigDecimal) ? EvalHelper.getBigDecimalOrNull( result.toString() ) : result;
+        if (result != null && result.getClass().isArray()) {
+            List<Object> objs = new ArrayList<>();
+            for (int i = 0; i < Array.getLength(result); i++) {
+                objs.add(EvalHelper.coerceNumber(Array.get(result, i)));
+            }
+            return objs;
+        } else {
+            return EvalHelper.coerceNumber(result);
```
Comment (pibizza) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java:372]: I assume there is a good structural reason to duplicate this code.

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNClassloaderTest.java [hunk 2]
```diff
@@ -127,4 +127,52 @@ public static String getPom(final ReleaseId releaseId, final ReleaseId... depend
...
+                                  "        return results;\n" +
+                                  "    }\n" +
+                                  "\n" +
+                                  "    public static int[] bkmI( String p1 ) {\n" +
+                                  "        int[] results = {9,8,7,6,5};" +
+                                  "        return results;\n" +
+                                  "    }\n" +
+                                  "}";
+
+        System.out.println(javaSource);
```
Comment (pibizza) [inline kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNClassloaderTest.java]: Why printing the source? I assume this is for debugging purpose, but once the test is running it should not be required.

**Review by pibizza — APPROVED** [review]

File: kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNClassloaderTest.java (hunk 2 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-core/src/test/java/org/kie/dmn/core/classloader/DMNClassloaderTest.java]: Just forgot it, nice catch thanks.
Removed with 6227e3175e.
I just used that for convenience to create related PR
 - https://github.com/kiegroup/droolsjbpm-integration/pull/2579

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java (hunk 1 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java:372]: Yes, the GWT transpilation (translation) of part of FEEL in some cases requires duplicate/adapter version of the base classes in `kie-dmn-feel`(Java) which is the source-of-truth.
This is one of these cases.

File: kie-dmn/kie-dmn-feel-gwt/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java [hunk 3]
```diff
@@ -592,7 +594,15 @@ public Object invoke(EvaluationContext ctx, Object[] params) {
...
-        return result != null && result instanceof Number && !(result instanceof BigDecimal) ? EvalHelper.getBigDecimalOrNull(result.toString()) : result;
+        if (result != null && result.getClass().isArray()) {
+            List<Object> objs = new ArrayList<>();
+            for (int i = 0; i < Array.getLength(result); i++) {
+                objs.add(EvalHelper.coerceNumber(Array.get(result, i)));
+            }
+            return objs;
+        } else {
+            return EvalHelper.coerceNumber(result);
+        }
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-feel-gwt/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java:605]: FYI @Rikkola @karreiro another instance of duplicated code in GWT that needed manual copy-pasta
/cc @danielezonca

Comment (tarilabs) [reply kie-dmn/kie-dmn-feel-gwt/src/main/java/org/kie/dmn/feel/runtime/functions/BaseFEELFunction.java:605]: (with reference to https://github.com/kiegroup/drools/pull/3798#discussion_r696555460 )

**Review by hellowdan — APPROVED** [review]

**Comment by tarilabs** [comment]
Jenkins run cdb

**Review by mariofusco — APPROVED** [review]

**Pull request merged**